### PR TITLE
Flag --output json and yaml added

### PR DESCRIPTION
**Pull request by deschmih**
refs #1534
refs #5588 

**What this PR does / why we need it**:
This pull request adds the flag --output json or yaml for repo_list, search, and install.
We want to use helm in our C# application. Currently we rely on parsing non-structured output which makes it hard to integrate in our software. Having the option to return helm output as JSON or YAML makes it easier to integrate helm in 3rd party applications that cannot use the Go API.

**Special notes for your reviewer**:
I did something wrong with the last pull request #5588, because I am a newbie with git ;-)

**If applicable**:
- [x ] this PR contains documentation
- [x ] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by asherf** [comment]
@deschmih any chance of rebasing this and getting it in ?

**Comment by deschmih** [comment]
@asherf : Conflicts are resolved.

**Comment by deschmih** [comment]
@thomastaylor312 Is it possilbe to go ahead an give this a review. This would be very nice.

**Comment by thomastaylor312** [comment]
@deschmih Right now, we are worried about accepting too many more features, because we are getting close to cutting 3.0 and all of the new features would need to be supported in Helm 3 (which the more we need to port, the more likely we are to cause bugs). On the other hand though, we already do have similar output flags in other places and it is likely to be useful. So we have a couple options here and I'd like your thoughts on them:

1. You could just port this PR directly to Helm 3 and have it be a new feature there.
2. We could pull this in to Helm 2 and if you have the time, _also_ port it to Helm 3

From a maintainer standpoint, it'd be much better to do option 1, but if that won't work for you, like I said, we have similar code elsewhere, so I don't consider this to be as huge of a worry to port over to Helm 3 after we merge

**Comment by deschmih** [comment]
@thomastaylor312 Thanks for your reply. At the moment it is important for me to have this feature in helm 3. So I will implement this PR in helm 3, if i can get you as a reviewer :-).

**Comment by thomastaylor312** [comment]
Sure thing @deschmih! I'll get this reviewed today and feel free to ping me on the v3 port

File: cmd/helm/repo_list.go
Comment (thomastaylor312) [inline cmd/helm/repo_list.go]: Much of this code in all of the printers is pretty repetitive and could be refactored out to be a bit more generic. I think to make this a little bit more maintainable long term, we should probably invest in doing that. For example, you could combine the for loops putting together the lists into a single case statement and then have a `PrintFormatted(obj interface{}, type string)` that can then marshal the data appropriately. That way we don't have duplicate error handling and marshaling everywhere. Another possibility would be to have all of these element/container types implement some sort of interface that has methods for printing a table, json, or yaml. In any case, there are multiple ways to solve this problem and I'd be open to any of them. Let me know if I wasn't clear or if you have any questions

**Review by thomastaylor312 — CHANGES_REQUESTED** [review]
Hey @deschmih thanks for all the hard work on this! So this looks like it delivers really well on all of the functionality. My main concern is maintainability, and so I think there should be a little bit of refactoring to make the printing a little bit cleaner and more DRY. See the specific code comments for more details

File: cmd/helm/repo_list.go
Comment (thomastaylor312) [reply cmd/helm/repo_list.go]: By the way, feel free to ping me on Slack if you have any questions or would like to pair

**Comment by thomastaylor312** [comment]
Hey @deschmih! Things are looking better! Once those tests are passing, I can do a final review and LGTM. There are a few things I think I'll want to refactor some more, but don't want to bog down this PR any longer for them

File: cmd/helm/status.go
Comment (thomastaylor312) [inline cmd/helm/status.go]: Looking at this again, I just realized this is a breaking change. This function was publicly exported, so it will need to remain that way, though the underlying logic can be shifted around

**Comment by thomastaylor312** [comment]
@deschmih I just opened #6381 with the refactor I had in mind that basically just reorganizes all of the great work you've done here. It also restores the unexported `PrintStatus` function so you don't have to worry about it here. 

I want to get this fixed and merged before the other PR so you get credit for the work you've put into this!

**Comment by deschmih** [comment]
@thomastaylor312 Thanks for your great work. 
I fixed my test and push it to my repository, but it fails again. During lifetime of this pull request I got his failure more than once and it's always:
**FAIL	k8s.io/helm/cmd/helm/installer [build failed]  
and/or
FAIL	k8s.io/helm/cmd/tiller [build failed]**

--> see pull/5593 9173 (build)
--> pull/5593 9443 (build)

I have no possibility to restart this test machine unless I make a other useless commit. I had discussed this topic at the helm-dev slack months ago, but the only advice was to make a push again.
Maybe a automatic restart, if this error occurs would be helpful?

https://kubernetes.slack.com/archives/C51E88VDG/p1554993109042300

**Comment by thomastaylor312** [comment]
@deschmih Yep, it all has to do with the Helm 2 builds hitting resource limits. I'll kick the build until we actually get clean signal. FWIW, it does not happen with Helm 3

**Comment by thomastaylor312** [comment]
@deschmih Looks like you need to `gofmt` `cmd/helm/upgrade.go`. It is failing the style tests. The command is `gofmt -s -w cmd/*` so you don't have to look it up 🙂

**Comment by deschmih** [comment]
@thomastaylor312  Thanks for your hint. I did the format job and pushed it again, but it still fails. :sob:

**Comment by thomastaylor312** [comment]
Me with the circle CI resource limits on Helm 2 builds right now: 
tenor-213758010

**Comment by karuppiah7890** [comment]
Omg. Lololol. CircleCI memes 🤣🤣🤣

**Comment by thomastaylor312** [comment]
More like "we somehow keep butting right up against the free tier limits between tiller and helm both being built"

In other news, @deschmih it passed!

**Review by thomastaylor312 — APPROVED** [review]

File: cmd/helm/upgrade.go
Comment (deschmih) [reply cmd/helm/upgrade.go]: That's because I did a merge-commit, it is always content of main. I had better done a rebase. Sorry.

Comment (thomastaylor312) [reply cmd/helm/upgrade.go]: @deschmih, let's get this added back in and rebase all the commits back down to one or two key commits

**Comment by deschmih** [comment]
@thomastaylor312 I did the rebase and squashed the commits, but surprise, surprise circle ci didn't like me. Please help.

**Review by bacongobbler — APPROVED** [review]

**Pull request merged**
